user/overriding-packages.rst

```diff
@@ -0,0 +1,280 @@
+Overriding Packages
+===================
+
+Assuming you know :doc:`what a workspace is <what-is-a-workspace>` and :doc:`how to use multiple workspaces<using-multiple-workspaces>`, you may decide you want to change the version of a package in an underlay without rebuilding the whole chain of workspaces.
+This can be done by  **overriding** the package.
+It is accomplished by sourcing an existing workspace, then building a different version of that package again in a new overlay workspace.
+
+Overriding a package is not always possible.
+This page offers tips for avoiding common issues.
+
+.. contents:: Table of Contents
+    :depth: 3
+
+
+How to avoid common issues
+--------------------------
+
+These are good practices to avoid common issues.
+If the advice is too restrictive then see the known issue descriptions for more complex advice.
+
+Use isolated workspaces
+***********************
+
+If you are building a workspace and suspect you may overriding a package from it in the future, then use an isolated workspace.
+This avoids include directory search order issues.
+See :doc:`this documentation <isolated-vs-merged-workspaces>` if you're unsure what an isolated workspace is.
+
+Override every package that depends on the one you want to override
+*******************************************************************
+
+A **leaf package** is one that has no other packages that depend on it.
+Overriding a non-leaf package can be problematic.
+Packages in the underlay were built against the underlay version of the package, but they will be expected to run with the overlay version.
+If their build process stores some information about the non-leaf package, such as an expected ABI, then undefined behavior can happen at runtime.
```

Comment:
Are you referring explicitly to "undefined behavior" in a C/C++ context or a more general "undefined behavior"? If the latter, I'd suggest saying "unexpected" or "unintended".

Runtime UB also isn't the only possible side effect as any build-time introspection would be affected. For example using build-time version info in a C/C++ preprocessor context to target different API versions of a dependency which is later overridden in an API-incompatible way.

Reply:
I was referring to more general undefined behavior. I used `unpredictable` instead in 6ae1e07